RFC-0007-remote-functions.md

```diff
@@ -0,0 +1,219 @@
+
+## Proposed Implementation
+
+### Design
+
+Fundamentally, the design of the REST API for remote functions will be based on the existing `FunctionNamespaceManager` interface SPI.
+Additionally, for Presto C++, currently there is no corresponding SPI for function namespace managers.  This RFC proposes to
+create a new REST-based implementation of the function execution framework.  By creating a REST API for remote functions that
+unifies the metadata and execution of remote functions, C++ deployments can similarly customize their function servers
+by implementing the REST API, in the same way that current Presto Java users can implement the `FunctionNamespaceManager` SPI.
+
+The REST API will power all `FunctionNamespaceManager` method implementations, including listing functions, retrieving function
+metadata, executing functions, and providing DDL support for functions (`ADD` and `DROP` support).
+
+#### Presto C++ special considerations
+
+The current Presto C++ implementation does not have a `FunctionNamespaceManager` SPI.  This RFC proposes to extend the Velox
+function server to support the REST API for remote functions.  This will allow for the definition of functions at runtime in
+Presto C++ deployments, and will unify function server implementations under one API.
+
+In the future, a plugin concept for remote functions should be added to Presto C++ to allow for custom function server implementations,
+however this is out of scope for this RFC.
+
+![REST API for remote functions](RFC-0007/Diagram.png)
+
+### Functions API
+
+An OpenAPI specification will be created which defines a REST API for remote function servers.  This API will feature the following
+endpoints:
+
+#### Functions version
+
+> Endpoint: /v1/functions
+>
+> HTTP verb: HEAD
+>
+> Request body: empty
+>
+> Response body: empty
+
+Returns the headers of the GET response.  This is useful for checking the version of the API, which will be returned as a header.
+
+#### List all functions
+
+> Endpoint: /v1/functions
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of functions across all schemas.
+
+#### List functions at schema
+
+> Endpoint: /v1/functions/{schema}
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of all functions in the specified schema.
+
+#### List functions with at schema with name
+
+> Endpoint: /v1/functions/{schema}/{functionName}
+>
+> HTTP verb: GET
+>
+> Request body: empty
+>
+> Response body: JSON array of function metadata objects
+
+Returns the complete listing of functions in the specified schema with the specified function name.
+
+#### Add a function
+
+> Endpoint: /v1/functions/{schema}/{functionName}
+>
+> HTTP verb: POST
+>
+> Request body: JSON object representing the function to be added
+>
+> Response body: the function ID of the newly created function
```

Comment:
Would be great to give more details about functionId. functionId is local to the function server. Will Presto use it at all ?